### DISCUSSION TITLE
How can I hide a field in the api?

### DISCUSSION THREAD

**Discussion by EduardoSpek** [pocketbase/pocketbase#138] [Q&A]
Hello guys! 
I'm From Brazil.

I'm using Google Translate now as I don't speak English.

I would like to know how I can hide fields in the api. Is it currently possible?

**Comment by ganigeorgiev** [comment]
Technically, you can if you prefix your field name with `#`. This is already implemented server side - https://github.com/pocketbase/pocketbase/blob/master/models/record.go#L291-L304.

But because this could cause some confusions in the Admin UI, currently you cannot type a field name from the interface with a `#` in its name.

I may reconsider allowing it in future versions, but will have to think a little more on it before reintroducing it.

If you really need to hide a certain API field, then you could use the collection create/update REST API and set the field name with it - https://pocketbase.io/docs/api-collections/ (requires `Authorization: Admin token` header to be set).

---

**Update:** The above is outdated and is no longer valid for v0.8. If you want to hide API fields, for now the easiest way would be to move the fields to a different collection and link them with a `relation` field in your main collection.

**Reply by ganigeorgiev** [reply]
For easier maintenance, I'd recommend to move your hidden fields in a different collection with its own API rules and in their place add a relation field pointing to the new collection with the hidden fields.

The end users will see only the id of the related records, but they'll not have access to them because of the API rules.

**Reply by EduardoSpek** [reply]
Thanks everyone for the informations. PocketBase is an amazing project. Congratulations! 👊🏽😎🇧🇷

**Reply by MeharGaur** [reply]
If you need to let the user *toggle* which fields are hidden or not, you can create a multi-select or JSON field called `private_fields`. Then, simply add an event hook on the `OnRecordViewRequest` and `OnRecordsListRequest` that modifies the response to strip out any fields that are in `private_fields`. So it's kind of like a middleware but it modifies the response, not the request. 

This is in theory, I haven't tested it yet but will be writing it out soon. Let me know if there's a better way to go about it.

**Reply by ganigeorgiev** [reply]
@MeharGaur Yes, this could be another way to do it, but this also implies that you are using PocketBase as framework. Additionally, although the fields can be hidden from the response this way, you may need to take care also for stripping them from the  `filter` and `sort` query parameters (which  may not be so easy as it sounds).

**Reply by YoRHa-A5** [reply]
@ganigeorgiev Any updates on this? The ability to hide fields from the api response would be very useful, exactly what I'm looking for rn :D
From my own code when trying to build a custom backend I remember doing it like this:

```
type User struct {
	ID           primitive.ObjectID   `json:"-" bson:"_id"`
	Name         string               `json:"username" bson:"username"`
	Email        string               `json:"email" bson:"email"`
	Password     password             `json:"-" bson:"password"`
}
```
where `json:"-"` was automatically hiding the info from the response. Maybe this helps?

**Reply by ganigeorgiev** [reply]
@BunnyTheLifeguard I don't have updates on this. Other tasks are currently with higher priority. I still recommend if you want to hide fields from the response to move them to a separate collection with its own API rules. This way you'll have more control on who can view/set/modify them.

**Reply by YoRHa-A5** [reply]
@ganigeorgiev I see. I guess I have to use a different collection `publicUser` or sth like that since displaying a user profile for another user would provide access to their email address etc. have to think about how I structure the data. 🤔 Anyway thanks man 👍🏻

**Reply by ganigeorgiev** [reply]
> since displaying a user profile for another user would provide access to their email address etc.

Since PocketBase v0.8, each auth record has `emailVisibility` field that users can use to control whether they want their email to be public or not.

**Comment by MarvinJWendt** [comment]
I use this hook to whitelist specific fields:

```go
	app.OnRecordViewRequest().Add(func(e *core.RecordViewEvent) error {
		if e.Collection.Name != "YOUR_COLLECTION_NAME" {
			return nil
		}

		allowedFields := []string{"id", "name", "created"}
		newMap := make(map[string]interface{}, len(allowedFields))

		for _, field := range allowedFields {
			newMap[field] = e.Record.Get(field)
		}

		err := e.HttpContext.JSON(200, newMap)
		if err != nil {
			return err
		}

		// Prevent the default response
		e.HttpContext = nil

		return nil
	})
```

I am not sure tho, if `e.HttpContext = nil` has any unintended side effects, such as having a reader or writer open.

**Reply by ganigeorgiev** [reply]
I'm not sure how actually this even works (I'd expect to throw a panic since the hook finalizer is using `e.HttpContext`).

If you want to prevent the default response handler, you can return `hook.StopPropagation` from inside the hook (_in future versions this will not be necessary and the response body write will be skipped if the response was already "committed"_).

**Reply by ganigeorgiev** [reply]
The above probably works because the panic is handled by the recovery middleware and the fact the we ignore the panic->error response handler if the response was already commited - https://github.com/pocketbase/pocketbase/blob/master/apis/base.go#L49 (I'll consider adding at least a dev debug log).

But the recommended way to prevent the default response handler is to return `hook.StopPropagation`.

**Comment by youssefsiam38** [comment]
I used this hook to do this in OnRecordsListRequest

```go
	app.OnRecordsListRequest("accounts").Add(func(e *core.RecordsListEvent) error {
		admin := e.HttpContext.Get(apis.ContextAdminKey)
		if admin != nil {
			return nil
		}
		body := make(map[string]any)
		newMap := []map[string]any{}

		// remove the field "token" from the response
		for _, field := range e.Records {

			newField := map[string]any{}
			for k, v := range field.ColumnValueMap() {
				if k == "token" {
					continue
				}
				newField[k] = v

			}

			newMap = append(newMap, newField)
		}

		body = map[string]any{
			"page":       e.Result.Page,
			"perPage":    e.Result.PerPage,
			"totalItems": e.Result.TotalItems,
			"totalPages": e.Result.TotalPages,
			"items":      newMap,
		}

		err := e.HttpContext.JSON(200, body)
		if err != nil {
			return err
		}

		return err
	})
```

**Reply by ganigeorgiev** [reply]
@youssefsiam38 This is a very old discussion and a better approach for hiding or returning only a subset of fields would be to use the [View collection](https://pocketbase.io/docs/collections/#view-collection).

With the [planned refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) there will be a "OnSerialize" hook that will allow modifying the record before serialization.

**Comment by AnatoleLucet** [comment]
In case this might help someone:
```golang
app.OnRecordViewRequest("my_collection").Add(func(e *core.RecordViewEvent) error {
  res := e.Record.PublicExport()
  delete(res, "private_field")

  return e.HttpContext.JSON(200, res)
})

app.OnRecordsListRequest("my_collection").Add(func(e *core.RecordsListEvent) error {
  items := make([]map[string]any, 0, len(e.Records))

  for _, r := range e.Records {
    r := r.PublicExport()
    delete(r, "private_field")

    items = append(items, r)
  }

  e.Result.Items = items
  return e.HttpContext.JSON(200, e.Result)
})

// and the same could probably be done with OnRealtimeBeforeMessageSend
```

I couldn't use a View collection for my specific use case but it's probably the way to go. Creating a sub-collection for private fields might be a good option too (that's what i ended up doing).
The above is probably a bit hacky and might break in future versions, but if you really need it, there it is.

**Reply by ganigeorgiev** [reply]
_Side-note: I don't recommend doing the above as it is very error prone. Instead, with the upcoming v0.23.0 release you can make use of the `OnRecordEnrich` hook as shown in https://github.com/pocketbase/pocketbase/discussions/5369#discussioncomment-10524248._
